python/jiminy_py/src/jiminy_py/dynamics.py

```diff
@@ -232,59 +229,235 @@ def get_body_world_acceleration(robot, body_name, use_theoretical_model=True):
 
     return spatial_acceleration
 
-def compute_freeflyer_state_from_fixed_body(robot, fixed_body_name, position,
-                                            velocity=None, acceleration=None,
+def _compute_closest_contact_frame(robot, ground_profile=None, use_theoretical_model=True):
+    """
+    @brief   Compute the closest contact point to the ground, in their respective local
+             frame and wrt the ground position and orientation.
+
+    @details This method can be used in conjunction with compute_freeflyer_state_from_fixed_body
+             to determine the right fixed_body_name that ensures no contact points are going
+             through the ground and a single one is touching it.
+
+             It is assumed that update_quantities has been called.
+
+    @param robot     The jiminy robot
+    @param ground_profile   The ground profile callback
+    @param position   Joint position vector
+    """
+
+    if use_theoretical_model:
+        pnc_data = robot.pinocchio_data_th
+    else:
+        pnc_data = robot.pinocchio_data
+
+    # Compute the transform in the world of the contact points
+    contact_frames_transform = []
+    for frame_idx in robot.contact_frames_idx:
+        transform = pnc_data.oMf[frame_idx]
+        contact_frames_transform.append(transform)
+
+    # Compute the transform of the ground at these points
+    if ground_profile is not None:
+        contact_ground_transform = []
+        ground_pos, ground_rot = np.zeros(3), np.empty((3, 3))
+        for frame_transform in contact_frames_transform:
+            ground_pos[2], ground_rot[2] = ground_profile(frame_transform.translation)
+            ground_rot[1] = np.cross(ground_rot[2], np.array([1.0, 0.0, 0.0]))
+            ground_rot[0] = np.cross(ground_rot[1], ground_rot[2])
+            contact_ground_transform.append(pin.SE3(ground_rot, ground_pos))
+    else:
+        contact_ground_transform = len(contact_frames_transform) * [pin.SE3.Identity()]
+
+    # Compute the position and normal of the contact points wrt their respective ground transform
+    contact_frames_pos_rel = []
+    contact_frames_normal_rel = []
+    for frame_transform, ground_transform in \
+            zip(contact_frames_transform, contact_ground_transform):
+        transform_rel = ground_transform.actInv(frame_transform)
+        contact_frames_pos_rel.append(transform_rel.translation)
+        contact_frames_normal_rel.append(transform_rel.rotation[:, 2])
+
+    # Compute the closest contact points to the ground in their respective local frame
+    for i in range(len(robot.contact_frames_idx)):
+        height_frame = contact_frames_pos_rel[i] @ contact_frames_normal_rel[i]
+        is_closest = True
+        for j in range(i+1, len(robot.contact_frames_idx)):
+            height = contact_frames_pos_rel[j] @ contact_frames_normal_rel[i]
```

Comment:
hum... I don't think so since we don't know in advance which ground_transform will be applied at the end, we need to test them all.

Reply:
I didn't see the double loop ; now I don't understand what exactly you are doing.

I remember we discussed when working on the ground contact formula that finding the ground-point distance was difficult, but I don't remember how we solved it (I think it was through some heuristic)...

i guess the start is : what is `ground_pos[2], ground_rot[2] = ground_profile(frame_transform.translation)` doing ? I thought it was computed this closest point (given frame_transform.translation, it gives the point 'below' this point on the ground. Then you compute the height and take the minimum height ?

Reply:
To sum up, here is what I'm trying to do:
I want to find the contact point along the list that ensures 2 things:
-  no contact points are going through the ground 
- a single one is touching the ground
This contact point will be referred as the **"true contact point"** in the following.

To do so, I assume that a contact point is the *"true contact point"*, using this assumption, I compute the the height of every contact points, and then I check if the assumption holds true (If the 2 conditions I mentioned are valid). If not, i discard this contact point and try the next with the one.

Reply:
Ah I see, you need this because you are trying to align the norm of the contact body with the normal of the ground.

This is kind of confusing to me, because when you choose to do ```             ground_pos[2], ground_rot[2] = ground_profile(frame_transform.translation)``` you are doing a vertical projection of the contact point on the ground. This is really the simplest solution, as we already discussed, because finding the right point to do an orthogonal projection is difficult.But this is inconsistent with changing the rotation as if you were doing an orthogonal projection - which is what you are doing after. 

I think we need to choose clearly what this function is doing. If it's only to 'drop' the robot on the ground, lowering its height to touch the ground (but not changing any orientation), then the code is much simpler. If instead you want to really find the closest point between the ground and the contact point, and then assume that the foot is tangent to the ground at this point, then the code is more complexe.... but I'm not sure this is what you want either. 

What is really the aim of this function ? If it's just for display we can tolerate whatever sensible approximation...

Reply:
> you are trying to align the norm of the contact body with the normal of the ground.

Exactly

> But this is inconsistent with changing the rotation as if you were doing an orthogonal projection - which is what you are doing after.

True

> If instead you want to really find the closest point between the ground and the contact point, and then assume that the foot is tangent to the ground at this point

Yes, that's what I'm trying to do. It is true that it is not what we want, actually we should make sure that 3 contact points are touching the ground but I'm doing this approximation to be consistent with `retrieve_freeflyer`, which requires the name of a single frame of align with.

> What is really the aim of this function ? If it's just for display we can tolerate whatever sensible approximation...

This function is used as an intermediary step to estimate all the kinematics and dynamics data of a given trajectory based on the rigid joints evolution only. To be more specific, I would like to estimate the position of the center of pressure and the motor torques.

Reply:
ping @matthieuvigne  UP !